File: 프로그래머스/2020 KAKAO BLIND RECRUITMENT/괄호변환.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;
bool check(string p){
    stack<char> s;
    // true : 올바른 괄호 문자열
    for(auto c : p){
        if(c == '(') s.push(c);
        else if(!s.empty() && c == ')' && s.top() == '(') s.pop();
        else return false;
    }
    return true;
}

string parsing(string s){
    string result = "";
    if(s == "") return s;
    int open = 0;
    int close = 0;
    int index = 0;

    for(auto c : s){
        if(c == '(') open ++;
        else close ++;
        index ++;
        if(open == close) break;
    }
    string u = s.substr(0,index);
    string v = s.substr(index);
    if(check(u)) {
        result += u;
        result += parsing(v);
    }
    else{
        result += "(";
        result += parsing(v);
        result += ")";
        u = u.substr(1,u.size()-2);
        for(auto a : u){
            if(a == '(') result += ")";
            else result += "(";
        }
    }
    return result;
}

string solution(string p) {
    if(check(p)) return p;
    return parsing(p);
}

File: 프로그래머스/2020 KAKAO BLIND RECRUITMENT/괄호변환.js
// 괄호변환
const divide = function(p){
    let open = 0, close = 0;
    for(let i = 0; i < p.length; i++){
        if(p[i] === '(') open += 1;
        else close += 1;
        if(open === close) return i
    }
}
const check = function(p){
    let open = 0;
    p = p.split("");
    p.forEach((ele) => {
        if(ele === '(') open += 1;
        else{
            if(open === 0) return false;
            else open -= 1;
        }
    });
    return open === 0; 
}
function solution(p) {
    let answer = '';
    if (p.length === 0) return p;
    const idx = divide(p);
    const u = p.slice(0,idx+1);
    const v = p.slice(idx+1);
    if(check(u)) {
        answer += u;
        answer += solution(v);
    }
    else {
        answer += '(';
        answer += solution(v);
        answer += ')';
        if(u.length > 2){
            const new_u = u.slice(1,u.length-1).split("");
            new_u.forEach((ele) => {
                if(ele === '(') answer += ')';
                else answer += '(';
            });
        }
    }
    return answer;
}